File: node/src/main/java/dev/httpmarco/polocloud/node/terminal/commands/ServiceCommand.java
package dev.httpmarco.polocloud.node.terminal.commands;

import dev.httpmarco.polocloud.node.Node;
import dev.httpmarco.polocloud.node.commands.Command;
import dev.httpmarco.polocloud.node.commands.CommandArgumentType;
import lombok.extern.log4j.Log4j2;

@Log4j2
public final class ServiceCommand extends Command {

    public ServiceCommand() {
        super("service", "Manage all your services", "services", "ser");

        var serviceArgument = CommandArgumentType.ClusterService("service");

        syntax(it -> {
            var serviceProvider = Node.instance().serviceProvider();

            log.info("Following &b{} &7groups are loading&8:", serviceProvider.services().size());
            serviceProvider.services().forEach(group -> log.info("&8- &f{}&8: (&7{}&8)", group.name(), group.details()));
        }, CommandArgumentType.Keyword("list"));

        syntax(it -> {
            var service = it.arg(serviceArgument);
            log.info("Name&8: &b{}", service.name());
            log.info("Running node&8: &b{}", String.join("&8, &b", service.runningNode()));
            log.info("Group&8: &b{}", service.group().name());
            log.info("Bind&8: &b{}&8:&b{}", service.hostname(), service.port());
            log.info("State&8: &b{}", service.state());
            log.info("Maximal players: &b{}" , service.maxPlayers());
            log.info("Online players&8: &b{}", service.onlinePlayersCount());
            log.info("Properties&8(&b{}&8):", service.properties().properties().size());
            service.properties().properties().forEach((property, o) -> log.info("&8 - &7{}&8 = &7{}", property.name(), o));
        }, serviceArgument, CommandArgumentType.Keyword("info"));


        syntax(it -> {
            var service = it.arg(serviceArgument);
            log.info("Trying to shutdown &8'&f{}'&8", service.name());

            // todo
            service.shutdown();
        }, serviceArgument, CommandArgumentType.Keyword("shutdown"));


        syntax(it -> {
            var service = it.arg(serviceArgument);
            Node.instance().terminal().clear();
            log.info("Logs of service {}&8: ", service.name());

            for (var logLine : service.logs()) {
                log.info(logLine);
            }
        }, serviceArgument, CommandArgumentType.Keyword("log"));

        syntax(it -> Node.instance().screenProvider().display(it.arg(serviceArgument)), serviceArgument, CommandArgumentType.Keyword("screen"));

        var commandArg = CommandArgumentType.StringArray("command");
        syntax(it -> {
            var service = it.arg(serviceArgument);
            var command = it.arg(commandArg);

            service.executeCommand(command);
            log.info("&b{} &8-> &b{} &8| &7{}", Node.instance().clusterProvider().localNode().data().name(), service.name(), command);
        }, serviceArgument, CommandArgumentType.Keyword("execute"), commandArg);
    }
}

File: agent/src/main/java/dev/httpmarco/polocloud/agent/runtime/local/terminal/commands/impl/ServiceCommand.kt
package dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.impl

import dev.httpmarco.polocloud.agent.Agent
import dev.httpmarco.polocloud.agent.logger
import dev.httpmarco.polocloud.agent.runtime.RuntimeServiceStorage
import dev.httpmarco.polocloud.agent.runtime.local.terminal.Jline3Terminal
import dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.Command
import dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.type.KeywordArgument
import dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.type.ServiceArgument
import dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.type.StringArrayArgument
import kotlin.collections.component1
import kotlin.collections.component2

class ServiceCommand(private val serviceStorage: RuntimeServiceStorage, terminal: Jline3Terminal) :
    Command("service", "Used to manage services", "ser") {

    init {
        syntax(execution = { context ->
            if (serviceStorage.items().isEmpty()) {
                logger.info("No service found.")
                return@syntax
            }
            logger.info("Found ${serviceStorage.items().size} services&8:")
            serviceStorage.items().forEach { logger.info(" &8- &3${it.name()} &8(&7&8)") }
        }, KeywordArgument("list"))

        var serviceArgument = ServiceArgument()

        syntax(execution = {
            var service = it.arg(serviceArgument)

            logger.info("Service &3${service.name()}&8:")
            logger.info(" &8- &7State&8: &f${service.state}")
            logger.info(" &8- &7Group&8: &f${service.group.data.name}")
            logger.info(" &8- &7Port&8: &f${service.port}")
            logger.info(" &8- &7Hostname&8: &f${service.hostname}")
            logger.info(" &8- &7Players&8: &f${service.playerCount}&8/&f${service.maxPlayerCount}")
            logger.info(" &8- &7Properties&8:")

            service.properties.forEach { (key, value) ->
                logger.info("   &8- &7$key&8: &f$value")
            }
        }, serviceArgument)

        syntax(execution = {
            var service = it.arg(serviceArgument)

            service.shutdown()
        }, serviceArgument, KeywordArgument("shutdown"))

        syntax(execution = {
            var service = it.arg(serviceArgument)

            service.logs().forEach { log ->
                println(log)
            }
        }, serviceArgument, KeywordArgument("logs"))

        var commandArg = StringArrayArgument("command")
        syntax(execution = {
            var service = it.arg(serviceArgument)

            Agent.instance.runtime.expender().executeCommand(service, it.arg(commandArg))
            logger.info("Executed command on service &3${service.name()}&8: &f${it.arg(commandArg)}")
        }, serviceArgument, KeywordArgument("execute"), commandArg)

        syntax(execution = {
            val service = it.arg(serviceArgument)
            terminal.screenService.screenRecordingOf(service)
        }, serviceArgument, KeywordArgument("screen"))
    }
}